CRM/Core/DAO/AllCoreTables.php

```diff
@@ -73,33 +68,71 @@ public static function flush(): void {
    * @param string $links_callback
    */
   public static function registerEntityType($briefName, $className, $tableName, $fields_callback = NULL, $links_callback = NULL) {
-    self::$daoToClass[$briefName] = $className;
-    self::$tables[$tableName] = $className;
-    self::$entityTypes[$briefName] = [
-      'name' => $briefName,
+    Civi::$statics[__CLASS__]['tables'][$tableName] = $briefName;
+    Civi::$statics[__CLASS__]['classes'][$className] = $briefName;
+    Civi::$statics[__CLASS__]['entities'][$briefName] = [
       'class' => $className,
       'table' => $tableName,
-      'fields_callback' => $fields_callback,
-      'links_callback' => $links_callback,
     ];
+    if ($fields_callback) {
+      Civi::$statics[__CLASS__]['entities'][$briefName]['fields_callback'] = $fields_callback;
+    }
+    if ($links_callback) {
+      Civi::$statics[__CLASS__]['entities'][$briefName]['links_callback'] = $links_callback;
+    }
+  }
+
+  /**
+   * @return array[]
+   *   [EntityName => [table => table_name, class => CRM_Core_DAO_EntityName]][]
+   */
+  public static function getEntities(): array {
+    self::init();
+    return Civi::$statics[__CLASS__]['entities'];
+  }
+
+  /**
+   * @return string[]
+   *   [table_name => EntityName][]
+   */
+  private static function getEntitiesByTable(): array {
+    self::init();
+    return Civi::$statics[__CLASS__]['tables'];
   }
 
   /**
+   * @return string[]
+   *   [CRM_Core_DAO_ClassName => EntityName][]
+   */
+  private static function getEntitiesByClass(): array {
+    self::init();
+    return Civi::$statics[__CLASS__]['classes'];
+  }
+
+  /**
+   * @deprecated in 5.72 will be removed in 5.90.
```

Comment:
@totten in order to cut down memory use I removed 3 items from the cached array: 'name' (redundant with array key), 'fields_callback' & 'links_callback' (if NULL):

**Before:**
```
'Contact' => [
  'name' => 'Contact',
  'class' => 'CRM_Contact_DAO_Contact',
  'table' => 'civicrm_contact',
  'fields_callback' => NULL,
  'links_callback' => NULL,
]
```

**After:**
```
'Contact' => [
  'class' => 'CRM_Contact_DAO_Contact',
  'table' => 'civicrm_contact',
]
```

So I had a few choices about how manage this change in the `get()` function:
1. Just change the signature without notice & hope nobody cares. (risky?)
2. Every time the `get()` function is called, loop through the array and add them back before returning (slow).
3. Create a new function and deprecate the existing one, doing the loop to preserve output format, which is slow but unused in core.

I went with option 3.

Reply:
@totten aside, what are `fields_callback` and `links_callback` used for? I don't see any reference to them in core outside of [a unit test](https://github.com/civicrm/civicrm-core/blob/9df1541a8e0c1520f1879ea75cf171b13af08a91/tests/phpunit/CRM/Core/DAO/AllCoreTablesTest.php#L44-L49) and there's no explanation in your [commit message](https://github.com/civicrm/civicrm-core/pull/8769/commits/740dd877cde870c7a80d7e0d1b143e5dee8fb37f), nor do I see any rationale given in the [linked issue](https://issues.civicrm.org/jira/browse/CRM-19130).

Reply:
Yeah, I can see how the new method is the safer way to change the signature.

The [memory-usage of a repeated string](https://gist.github.com/totten/befe1a3be03f4c24df99e4addd4658ed) is interesting. If you have 10x records with the `fields_callback` string, that doesn't mean that there are 10x copies of `f` `i` `e` `l` `d` `s` `_` `c` `a` `l` `l` `b` `a` `c` `k`. There's one copy with 10x pointers. (Well, it probably depends on the exact origin the string. Two file-buffers will be two file-buffers.  But in this case, all copies originate from the same literal expression/bytecode. If the interpreter is sane, then it's just one copy.)

Where it might matter is if you're angling to move the cache to an external data-store (redis/memcache). Neither JSON nor PHP-serialize are optimized for that kind of thing.

Reply:
> nor do I see any rationale given in the [linked issue](https://issues.civicrm.org/jira/browse/CRM-19130) other than "it would be nice to have".

Doesn't everything boil down that. 😜  But more seriously, they talk about altering properties like import/export/pattern-detection... which is fiddly-heuristic stuff; in a pinch ("You have 60 sites which all have similar import workflows and make the same mistakes in building their import mappings... and you don't want to spend 4 months redesigning the whole import system..."), I think that's relateable.

>  > Where it might matter is if you're angling to move the cache to an external data-store
>
> Yea that's where this is headed...

Right, I don't want to prejudge the encoding. For deciding if it matters, benchmarks are truth... But I can see a sort of appeal in "if it's empty, then you might as well drop it".

> I see how that would work with array values, but what about array keys? 

Well, the same theory would make sense IMHO. Keys and values are both scalars, and they're both payloads for the array. But it's a good question, because  PHP could have some subtle variations.

https://gist.github.com/totten/4c73995bde21e16a22d97706cfb7aeb0

This defines the key in different ways. Notably, the control-case(`#0` with `$key=''`) has the same memory-usage as the cases with long keys. (With inefficient representations, you would expect the long keys to use an extra 200kb.) The `$array` is just a graph of pointers. It doesn't matter the size of things that you're pointing at.. if they're all the same things...

Reply:
@totten all good to know. Happy to keep discussing but also... let's merge this PR!